### PR TITLE
DOC: fail to build if there are warnings

### LINKED ISSUE #38011: Build documentation without emitting warnings
Currently documentation builds emit many warnings. We should resolve them, then turn on `-WT --keep-going` to report warnings as errors so that new build problems are caught as they are created.

cc @ezyang @zou3519

### PR DESCRIPTION
**Pull request by mattip**
Merge after gh-41334 and gh-41321 (EDIT: both are merged).
Closes gh-38011

This is the last in a series of PRs to build documentation without warnings. It adds `-WT --keepgoing` to the shpinx build which will [fail the build if there are warnings](https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-W), print a [trackeback on error](https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-T) and [finish the build](https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-keep-going) even when there are warnings.

It should fail now, but pass once the PRs mentioned at the top are merged.

### CODE REVIEW

**Comment by mattip** [comment]
xref gh-32703: should this also close that issue?

**Review by rgommers — APPROVED** [review]

**Comment by rgommers** [comment]
Prerequisites are merged, so let's see if this is still passing.

**Comment by rgommers** [comment]
Hmm the bot seems to be asleep. Also, the `pytorch_doc_test` job shows some warnings that cause a CI failure. @mattip can you address those and rebase?

File: docs/Makefile [hunk 1]
```diff
@@ -2,7 +2,7 @@
 #
 
 # You can set these variables from the command line.
-SPHINXOPTS    ?= -j auto
+SPHINXOPTS    ?= -j auto -WT --keep-going
```
Comment (mattip) [inline docs/Makefile:5]: This is the actual one-liner meant to be in the PR

File: docs/source/jit.rst [hunk 2]
```diff
@@ -50,7 +50,6 @@ Creating TorchScript Code
     wait
     ScriptModule
     ScriptFunction
-    freeze
```
Comment (mattip) [inline docs/source/jit.rst:53]: freeze -> _freeze, it is now a private module

File: torch/__init__.py [hunk 3]
```diff
@@ -272,6 +272,7 @@ def set_default_tensor_type(t):
 def set_default_dtype(d):
     r"""Sets the default floating point dtype to :attr:`d`.
     This dtype is:
+
```
Comment (mattip) [inline torch/__init__.py:275]: blank line needed

File: torch/__init__.py [hunk 4]
```diff
@@ -282,7 +283,7 @@ def set_default_dtype(d):
     Args:
         d (:class:`torch.dtype`): the floating point dtype to make the default
 
-    Example::
+    Example:
```
Comment (mattip) [inline torch/__init__.py:286]: not really an error, just a cleanup

File: torch/functional.py [hunk 5]
```diff
@@ -1063,7 +1063,8 @@ def atleast_3d(*tensors):
     Returns:
         output (Tensor or tuple of Tensors)
 
-    Example::
+    Example:
+
```
Comment (mattip) [inline torch/functional.py:1067]: blank line preferred

File: torch/functional.py [hunk 6]
```diff
@@ -1076,7 +1077,7 @@ def atleast_3d(*tensors):
         >>> torch.atleast_3d(y)
         tensor([[[-0.8079],
                 [ 0.7460]],
-
+                <BLANKLINE>
```
Comment (mattip) [inline torch/functional.py:1080]: doctests need this to pass

File: torch/serialization.py [hunk 7]
```diff
@@ -334,7 +334,7 @@ def save(obj, f: Union[str, os.PathLike, BinaryIO],
          pickle_module=pickle, pickle_protocol=DEFAULT_PROTOCOL, _use_new_zipfile_serialization=True) -> None:
     """Saves an object to a disk file.
 
-    See also: :ref:`recommend-saving-models`
+    See also: `saving-loading-tensors`
```
Comment (mattip) [inline torch/serialization.py:337]: The file `docs/source/notes/serialization.rst` was refactored

**Review by mattip — COMMENTED** [review]
Most of these changes were needed since more changes to documentation after PRs gh-41334 and gh-41321 were made.

File: .circleci/scripts/python_doc_push_script.sh [hunk 8]
```diff
@@ -50,21 +50,17 @@ export PATH=/opt/conda/bin:$PATH
 
 rm -rf pytorch || true
 
-# Install TensorBoard in python 3 so torch.utils.tensorboard classes render
-pip install -q https://s3.amazonaws.com/ossci-linux/wheels/tensorboard-1.14.0a0-py3-none-any.whl
-
```
Comment (mattip) [inline .circleci/scripts/python_doc_push_script.sh:55]: this is now handled via requirements.txt on line 62

File: .circleci/scripts/python_doc_push_script.sh [hunk 9]
```diff
@@ -50,21 +50,17 @@ export PATH=/opt/conda/bin:$PATH
...
 # Get all the documentation sources, put them in one place
 pushd "$pt_checkout"
-checkout_install_torchvision
+# Do not do this, since torchvision is a separate project.
+# checkout_install_torchvision
 pushd docs
-rm -rf source/torchvision
-cp -a ../vision/docs/source source/torchvision
+# rm -rf source/torchvision
+# cp -a ../vision/docs/source source/torchvision
```
Comment (mattip) [inline .circleci/scripts/python_doc_push_script.sh:59]: torchvision is building its own documentation at https://pytorch.org/vision. This was fixed in gh-41334

**Review by mattip — COMMENTED** [review]
Fixed the docs build script too.

File: torch/functional.py (hunk 6 above)
Comment (rgommers) [reply torch/functional.py:1080]: There are zero other instances of `BLANKLINE`, so I guess that makes this the only 3-D array in an example. Hence not worth cleaning this up with better doctest machinery I think.

A bit odd though that `doctest` failures don't turn into actual failures, that should be separate from `-WT` right?

File: .circleci/scripts/python_doc_push_script.sh (hunk 9 above)
Comment (rgommers) [reply .circleci/scripts/python_doc_push_script.sh:59]: Do you need to keep those lines as comments rather than just deleting them?

**Review by rgommers — APPROVED** [review]

File: .circleci/scripts/python_doc_push_script.sh (hunk 9 above)
Comment (mattip) [reply .circleci/scripts/python_doc_push_script.sh:59]: removing

**Comment by rgommers** [comment]
LGTM now, doc build all green. @jlin27 or @mruberry can one of you land this?

**Review by mruberry — APPROVED** [review]
Thanks @mattip!

**Pull request closed without merging**